components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java

```diff
@@ -281,6 +288,188 @@ private void handleRequest(HttpServletRequest req, HttpServletResponse resp, boo
+
+        log.debug("Invalid request message or single logout message ");
+
+        if (sessionId == null) {
+            String errorResp = SAMLSSOUtil.buildErrorResponse(
+                    SAMLSSOConstants.StatusCodes.REQUESTOR_ERROR,
+                    "Invalid request message", null);
+            sendNotification(errorResp, SAMLSSOConstants.Notification.INVALID_MESSAGE_STATUS,
+                    SAMLSSOConstants.Notification.INVALID_MESSAGE_MESSAGE, null, req, resp);
+        } else {
+            // Non-SAML request are assumed to be logout requests
+            sendToFrameworkForLogout(req, resp, null, null, sessionId, true,
+                    false);
+        }
+    }
+
+    private void handleSAMLResponse(HttpServletRequest req, HttpServletResponse resp, String samlResponse,
+                                    String sessionId, boolean isPost)
+            throws IdentityException, IOException, ServletException {
+
+        XMLObject response;
+
+        if (isPost) {
+            response = SAMLSSOUtil.unmarshall(SAMLSSOUtil.decodeForPost(samlResponse));
+        } else {
+            response = SAMLSSOUtil.unmarshall(SAMLSSOUtil.decode(samlResponse));
+        }
+
+        if (response instanceof LogoutResponse) {
+            LogoutResponse logoutResponse = (LogoutResponse) response;
+            handleLogoutResponseFromSP(req, resp, sessionId, logoutResponse);
+        } else {
+            handleInvalidRequestMessage(req, resp, sessionId);
+        }
+    }
+
+    private void handleLogoutResponseFromSP(HttpServletRequest req, HttpServletResponse resp, String sessionId,
+                                            LogoutResponse logoutResponse)
+            throws ServletException, IdentityException, IOException {
+
+        String inResponseToId = logoutResponse.getInResponseTo();
+        FrontChannelSLOParticipantInfo frontChannelSLOParticipantInfo =
+                getFrontChannelSLOParticipantInfo(inResponseToId);
+
+        if (frontChannelSLOParticipantInfo == null || !frontChannelSLOParticipantInfo.
+                getCurrentSLOInvokedParticipant().equals(logoutResponse.getIssuer().getValue())) {
+            handleInvalidRequestMessage(req, resp, sessionId);
+        } else {
+            // Remove front-channel SLO Participant info from the FrontChannelSLOParticipantStore.
+            removeFrontChannelSLOParticipantInfo(inResponseToId);
+            String logoutResponseIssuer = logoutResponse.getIssuer().getValue();
+            SAMLSSOServiceProviderDO responseIssuerSP = SAMLSSOUtil.getSPConfig(
+                    SAMLSSOUtil.getTenantDomainFromThreadLocal(), logoutResponseIssuer);
+
+            boolean isSuccessfullyLogout = SAMLSSOUtil.validateLogoutResponse(logoutResponse,
+                    responseIssuerSP.getCertAlias(), responseIssuerSP.getTenantDomain());
+
+            if (!isSuccessfullyLogout) {
+                // TODO : If the response is invalid, redirect the SP to an error page.
```

Comment:
Why TODO?

Reply:
As this PR was sent to a feature branch this TODO was added. This is addressed later in #208.